web/api/formatters/json_wrapper.c

```diff
@@ -85,115 +84,411 @@ static inline long jsonwrap_chart_ids(BUFFER *wb, const char *key, RRDR *r, RRDR
-        buffer_json_array_close(wb);
-    }
+static inline void query_target_data_statistics(BUFFER *wb, QUERY_TARGET *qt, struct query_data_statistics *d) {
+    buffer_json_member_add_object(wb, "sts");
+//    buffer_json_member_add_double(wb, "min", d->min);
+//    buffer_json_member_add_double(wb, "max", d->max);
+//    buffer_json_member_add_double(wb, "sum", d->sum);
+    buffer_json_member_add_double(wb, "avg", (d->count) ? d->sum / (NETDATA_DOUBLE)d->count : 0.0);
+//    buffer_json_member_add_double(wb, "vol", d->volume);
+    buffer_json_member_add_double(wb, "con", (qt->query_stats.volume > 0) ? d->volume * 100.0 / qt->query_stats.volume : 0.0);
+    buffer_json_object_close(wb);
+}
 
-    return 1;
+static void query_target_summary_hosts_v2(BUFFER *wb, QUERY_TARGET *qt, const char *key) {
+    buffer_json_member_add_array(wb, key);
+    for (long c = 0; c < (long) qt->hosts.used; c++) {
+        QUERY_HOST *qh = query_host(qt, c);
+        RRDHOST *host = qh->host;
+        buffer_json_add_array_item_object(wb);
+        buffer_json_member_add_string(wb, "mg", host->machine_guid);
+        if(qh->node_id[0])
+            buffer_json_member_add_string(wb, "nd", qh->node_id);
+        buffer_json_member_add_string(wb, "nm", rrdhost_hostname(host));
+        query_target_instance_counts(wb, &qh->instances);
+        query_target_metric_counts(wb, &qh->metrics);
+        query_target_alerts_counts(wb, &qh->alerts, NULL, false);
+        query_target_data_statistics(wb, qt, &qh->query_stats);
+        buffer_json_object_close(wb);
+    }
+    buffer_json_array_close(wb);
 }
 
-static inline void jsonwrap_full_dimension_list(BUFFER *wb, RRDR *r) {
-    QUERY_TARGET *qt = r->internal.qt;
+static size_t query_target_summary_contexts_v2(BUFFER *wb, QUERY_TARGET *qt, const char *key) {
+    buffer_json_member_add_array(wb, key);
+    DICTIONARY *dict = dictionary_create(DICT_OPTION_SINGLE_THREADED | DICT_OPTION_DONT_OVERWRITE_VALUE);
+
+    struct {
+        struct query_data_statistics query_stats;
+        struct query_instances_counts instances;
+        struct query_metrics_counts metrics;
+        struct query_alerts_counts alerts;
+    } x = {
+            .instances = (struct query_instances_counts){ 0 },
+            .metrics = (struct query_metrics_counts){ 0 },
+            .alerts = (struct query_alerts_counts){ 0 },
+    }, *z;
+
+    for (long c = 0; c < (long) qt->contexts.used; c++) {
+        QUERY_CONTEXT *qc = query_context(qt, c);
+
+        z = dictionary_set(dict, rrdcontext_acquired_id(qc->rca), &x, sizeof(x));
+
+        z->instances.selected += qc->instances.selected;
+        z->instances.excluded += qc->instances.selected;
 
+        z->metrics.selected += qc->metrics.selected;
+        z->metrics.excluded += qc->metrics.excluded;
+        z->metrics.queried += qc->metrics.queried;
+        z->metrics.failed += qc->metrics.failed;
+
+        z->alerts.clear += qc->alerts.clear;
```

Comment:
Hello @ktsaou ,

I started testing your PR to be sure we are not missing nothing, and when I ran coverity I got this report:

```sh
*** CID 436567:  Concurrent data access violations  (MISSING_LOCK)
/web/api/formatters/json_wrapper.c: 187 in query_target_summary_contexts_v2()
181             z->metrics.selected += qc->metrics.selected;
182             z->metrics.excluded += qc->metrics.excluded;
183             z->metrics.queried += qc->metrics.queried;
184             z->metrics.failed += qc->metrics.failed;
185     
186             z->alerts.clear += qc->alerts.clear;
>>>     CID 436567:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "z->alerts.warning" without holding lock "netdata_rwlock_t.rwlock_t". Elsewhere, "query_alerts_counts.warning" is accessed with "netdata_rwlock_t.rwlock_t" held 4 out of 5 times.
187             z->alerts.warning += qc->alerts.warning;
188             z->alerts.critical += qc->alerts.critical;
189     
190             query_target_merge_data_statistics(&z->query_stats, &qc->query_stats);
191         }
192    
```

Reply:
We have an analogous for other variables as well:

```sh
*** CID 436570:  Concurrent data access violations  (MISSING_LOCK)
/web/api/formatters/json_wrapper.c: 186 in query_target_summary_contexts_v2()
180     
181             z->metrics.selected += qc->metrics.selected;
182             z->metrics.excluded += qc->metrics.excluded;
183             z->metrics.queried += qc->metrics.queried;
184             z->metrics.failed += qc->metrics.failed;
185     
>>>     CID 436570:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "z->alerts.clear" without holding lock "netdata_rwlock_t.rwlock_t". Elsewhere, "query_alerts_counts.clear" is accessed with "netdata_rwlock_t.rwlock_t" held 4 out of 5 times.
186             z->alerts.clear += qc->alerts.clear;
187             z->alerts.warning += qc->alerts.warning;
188             z->alerts.critical += qc->alerts.critical;
189     
190             query_target_merge_data_statistics(&z->query_stats, &qc->query_stats);
191         }

** CID 436569:  Concurrent data access violations  (MISSING_LOCK)
/web/api/formatters/json_wrapper.c: 188 in query_target_summary_contexts_v2()
```